### DISCUSSION TITLE
Wrong type of variable in the class with `__slots__`

### DISCUSSION THREAD

**Discussion by belegnar** [microsoft/pyright#8742] [Q&A] [answered]
I believe it's a bug, but PR template suggests posting here first

What's wrong with `self.filters.pattern` here? `self.filters` is immutable, but even with
```
class Impl(Abstract):
    __slots__ = ()
    
    filters: RegExFilter = RegExFilter("(.+)")
```
`pyright` considers it mutable

```
class Filter:
    __slots__ = ()

    def do_filter(self):
        print("Main")

class RegExFilter(Filter):
    __slots__ = ("pattern",)

    def __init__(self, pattern: str):
        self.pattern = pattern

class Abstract(abc.ABC):
    __slots__ = ()

    filters: Filter

class Impl(Abstract):
    __slots__ = ()
    
    filters = RegExFilter("(.+)")

    def apply_filter(self):
        print(self.filters.pattern)
```

https://pyright-play.net/?code=JYWwDg9gTgLgBAQwEYGMBQaUBsEGddwBiwWMAplAFxpy1wD69uWEMujcAvHABQCUGOnAAmZAGYiI9MSXJQeuMljF9qQoWCjAAdjB4AiALIId%2BgZhz44AJTIBzAKIAPYqQo9Xc1TTqNmrdnouXn0wBBg5bX0AGnMhUQlGHWAYRgUlMWi4MIiKbUo4XBgob3U6RWUAOhzI4Jq8jGw8AgBBJCKoBBQ9ZBRKloAhAGFSuhk3KFwCzwpGywIASXAsHjaOrphR2nG5Am5bRxdZd30eSoBqPjNBOgTEMDAsAE9pY-kKlTUy7K1ddKqdhRcNVwpEBEA

**Comment by erictraut** [comment, marked as answer]
The fact that you're using `__slots__` here has nothing to do with the type violation you're seeing. You can completely remove `__slots__` from your example, and you'll see the same behavior.

You have declared that the variable `filters` in the class `Abstract` must be of type `Filter`. In the class `Impl`, you are assigning a value to this variable that indeed conforms to the type `Filter` (since `RegExFilter` is a subclass of `Filter`), so that is not a type violation. However, you're later relying on the fact that `Abstract.filters` is a `RegExFilter` rather than a `Filter`. That's a type violation, and pyright is correct to highlight it. There is nothing that prevents `Impl.filters` from being overwritten by another `Filter` instance — one that is not an instance of `RegExFilter`.

You could make your program type correct by adding an assert:

```python
    def apply_filter(self):
        assert isinstance(self.filters, RegExFilter)
        print(self.filters.pattern)
```

**Comment by belegnar** [comment]
`__slots__` makes sense here because with slots, `Impl.filters` is read-only and can't be anything other than a `RegExFilter`

**Reply by erictraut** [reply]
```python
class ImplSub(Impl): pass

i = ImplSub()
i.filters = Filter()
i.apply_filter() # Crash
```

**Reply by belegnar** [reply]
Indeed. Thx
